index.js

```diff
@@ -678,41 +681,22 @@
   };
 
   //  commonTypes :: ([[Type]], Boolean) -> [Type]
-  var commonTypes = function(typeses, loose) {
-    var types = chain(typeses, id);
-    if (isEmpty(types)) return [];
-
-    var $1s, $2s, candidates, idx, t;
-    if (types[0].type === 'UNARY') {
-      $1s = {};
-      for (idx = 0; idx < types.length; idx += 1) {
-        t = types[idx];
-        if (t.$1.type !== 'UNKNOWN') $1s[t.$1.name] = t.$1;
-      }
-      candidates =
-      map(or(map(keys($1s), function(k) { return $1s[k]; }), [Unknown]),
-          UnaryType.from(types[0]));
-    } else if (types[0].type === 'BINARY') {
-      $1s = {};
-      $2s = {};
-      for (idx = 0; idx < types.length; idx += 1) {
-        t = types[idx];
-        if (t.$1.type !== 'UNKNOWN') $1s[t.$1.name] = t.$1;
-        if (t.$2.type !== 'UNKNOWN') $2s[t.$2.name] = t.$2;
-      }
-      candidates = BinaryType.xprod(
-        types[0],
-        or(map(keys($1s), function(k) { return $1s[k]; }), [Unknown]),
-        or(map(keys($2s), function(k) { return $2s[k]; }), [Unknown])
-      );
-    } else {
-      candidates = types;
-    }
-
-    return filter(candidates, function(t1) {
-      return all(typeses, function(types) {
-        return any(types, function(t2) {
-          return equalTypes(t1, t2, loose);
+  //
+  //  [[String, RegexFlags], [String, RegexFlags]] ---> [String, RegexFlags]
+  //
+  //  [[Boolean], [Boolean], [Boolean], [Number]] ---> []
+  //
+  //  [[Array ???], [Array String], [Array ???]] ---> [Array String]
+  //
+  //  [[Either String ???], [Either ??? Number]] ---> [Either String Number]
+  var commonTypes = function($typeses, loose) {
+    return filter($typeses[0], function($t) {
+      return all($typeses, function(types) {
+        return any(types, function(t) {
+          var equal = equalTypes($t, t, loose);
+          if (equal && '$1' in $t && $t.$1.type === 'UNKNOWN') $t.$1 = t.$1;
+          if (equal && '$2' in $t && $t.$2.type === 'UNKNOWN') $t.$2 = t.$2;
```

Comment:
Wow! This is one of the best comments I have received on a pull request. Thank you, Aldwin.

> The caller seems to be aware of this [mutation], but doesn't seem to benefit from [it] in any way. (Correct me if I'm wrong)

You're correct. At one point I even had `$typeses.length = 0;` to prevent us from relying on `$typeses` after applying `customTypes`.

> Alternatively, I think you can stay away from mutation altogether by using `reduce` for your iterations, and then you'd be able to return the improved type info without mutating the input. If you want I can write the required reducers, because I [already kinda did](https://gist.github.com/Avaq/b15b787bf837e6094e0ba1804a75c5e0) to lay out my thoughts.

Wonderful! I'll gladly use this approach.

> I think you can do `tail` on `$typeses` before passing it to `all`. It would save you a single iteration, or in cases of a unary array, all of the work.

I thought so too. We have `NaN`-like values, though, such as `Array Inconsistent` (the type of `[Left('XXX'), 42]`). When `commonTypes` is applied to `[[Array Inconsistent]]` the result should be `[]` when `loose` is `false` or `[Array Inconsistent]` when `loose` is `true` (for displaying `Array ???`).

> By using `any` to iterate the types, you're not actually collecting all the type info you *could*.

Great point! I'll be sure to include a test for this.